### PR TITLE
Fix #2936: Remove deserialization hacks.

### LINKED ISSUE #2936: Remove deserialization hacks
* The actual *deserialization* hacks
* The link-time hack to generate bridges for inherited constructors

### PR DESCRIPTION
**Pull request by sjrd**
Remove both hacks actually happening during deserialization and those happening during the first linking step.

These changes are a prerequisite to implementing #2800 in the master branch.

### CODE REVIEW

**Comment by sjrd** [comment]
Jenkins, test this please

File: ir/src/main/scala/org/scalajs/core/ir/Serializers.scala [hunk 1]
```diff
@@ -690,8 +678,7 @@ object Serializers {
 
         case TagVarDef   => VarDef(readIdent(), readType(), readBoolean(), readTree())
         case TagParamDef =>
-          ParamDef(readIdent(), readType(), readBoolean(),
-              rest = if (useHacks060) false else readBoolean())
+          ParamDef(readIdent(), readType(), readBoolean(), readBoolean())
```
Comment (gzm0) [inline ir/src/main/scala/org/scalajs/core/ir/Serializers.scala]: Put on same line that `case` or space between `TagVarDef` and `TagParamDef`.

File: ir/src/main/scala/org/scalajs/core/ir/Serializers.scala [hunk 2]
```diff
@@ -739,12 +710,7 @@ object Serializers {
         case TagSelectStatic    => SelectStatic(readClassType(), readIdent())(readType())
         case TagApply           => Apply(readTree(), readIdent(), readTrees())(readType())
         case TagApplyStatically =>
-          val result1 =
-            ApplyStatically(readTree(), readClassType(), readIdent(), readTrees())(readType())
-          if (useHacks065 && result1.tpe != NoType && isConstructorName(result1.method.name))
-            result1.copy()(NoType)
-          else
-            result1
+          ApplyStatically(readTree(), readClassType(), readIdent(), readTrees())(readType())
```
Comment (gzm0) [inline ir/src/main/scala/org/scalajs/core/ir/Serializers.scala]: Same about spacing.

File: ir/src/main/scala/org/scalajs/core/ir/Serializers.scala [hunk 3]
```diff
@@ -798,90 +758,40 @@ object Serializers {
 
         case TagUndefinedParam => UndefinedParam()(readType())
 
-        case TagVarRef =>
-          val result = VarRef(readIdent())(readType())
-          if (useHacks060 && result.ident.name == "arguments")
-            foundArguments = true
-          result
+        case TagVarRef => VarRef(readIdent())(readType())
```
Comment (gzm0) [inline ir/src/main/scala/org/scalajs/core/ir/Serializers.scala]: Remove newlines and align arrows.

File: ir/src/main/scala/org/scalajs/core/ir/Serializers.scala [hunk 4]
```diff
@@ -1040,19 +939,13 @@ object Serializers {
...
-          case TagJSNativeLoadSpecImport =>
-            Some(JSNativeLoadSpec.Import(readString(), readStrings()))
-        }
+      (input.readByte(): @switch) match {
+        case TagJSNativeLoadSpecNone =>
+          None
+        case TagJSNativeLoadSpecGlobal =>
+          Some(JSNativeLoadSpec.Global(readStrings()))
+        case TagJSNativeLoadSpecImport =>
+          Some(JSNativeLoadSpec.Import(readString(), readStrings()))
```
Comment (gzm0) [inline ir/src/main/scala/org/scalajs/core/ir/Serializers.scala]: These needs newlines in between them.

Comment (sjrd) [inline ir/src/main/scala/org/scalajs/core/ir/Serializers.scala]: Why would those need newlines between them? They don't have blank lines inside their cases, so it's not required to have a blank line between the cases either.

Comment (gzm0) [inline ir/src/main/scala/org/scalajs/core/ir/Serializers.scala]: Ah, ok. My bad. I thought the style-guide said this. In that case you might want to optionally align the match for `readPropertyName`.

Comment (sjrd) [inline ir/src/main/scala/org/scalajs/core/ir/Serializers.scala]: Good point, done.

**Pull request merged**